esp-hal/src/uart/uhci/normal.rs

```diff
@@ -0,0 +1,371 @@
+
+        reg.conf0().modify(|_, w| w.len_eof_en().bit(config.len_eof));
+
+        if self.internal.set_chunk_limit(config.chunk_limit).is_err() {
+            return Err(AboveReadLimit);
+        }
+
+        Ok(())
+    }
+}
+
+// Based on SpiDmaTransfer
+/// A structure representing a DMA transfer for UHCI/UART.
+///
+/// This structure holds references to the UHCI instance, DMA buffers, and
+/// transfer status.
+#[instability::unstable]
+pub struct UhciDmaTxTransfer<'d, Dm, Buf>
+where
+    Dm: DriverMode,
+    Buf: DmaTxBuffer,
+{
+    uhci: ManuallyDrop<Uhci<'d, Dm>>,
+    dma_buf: ManuallyDrop<Buf::View>,
+}
+
+impl<'d, Buf: DmaTxBuffer> UhciDmaTxTransfer<'d, Async, Buf> {
+    fn new(uhci: Uhci<'d, Async>, dma_buf: Buf) -> Self {
+        Self {
+            uhci: ManuallyDrop::new(uhci),
+            dma_buf: ManuallyDrop::new(dma_buf.into_view()),
+        }
+    }
+
+    /// todo
+    pub fn is_done(&self) -> bool {
+        self.uhci.internal.channel.tx.is_done()
+    }
+
+    async fn wait_for_idle(&mut self) {
+        DmaTxFuture::new(&mut self.uhci.internal.channel.tx)
+            .await
+            .unwrap();
+    }
+
+    /// Waits for the DMA transfer to complete.
+    ///
+    /// This method blocks until the transfer is finished and returns the
+    /// `Uhci` instance and the associated buffer.
+    #[instability::unstable]
+    pub async fn wait(mut self) -> (Uhci<'d, Async>, Buf::View) {
+        self.wait_for_idle().await;
+
+        let retval = unsafe {
+            (
+                ManuallyDrop::take(&mut self.uhci),
+                ManuallyDrop::take(&mut self.dma_buf),
+            )
+        };
+        core::mem::forget(self);
+        retval
+    }
+
+    /// Cancels the DMA transfer.
+    #[instability::unstable]
+    pub fn cancel(mut self) -> (Uhci<'d, Async>, Buf::View) {
+        if !self.uhci.internal.channel.rx.is_done() {
+            self.uhci.internal.channel.rx.stop_transfer();
+        }
+
+        let retval = unsafe {
+            (
+                ManuallyDrop::take(&mut self.uhci),
+                ManuallyDrop::take(&mut self.dma_buf),
+            )
+        };
+        core::mem::forget(self);
+        retval
+    }
+}
+
+impl<Dm, Buf> Drop for UhciDmaTxTransfer<'_, Dm, Buf>
+where
+    Dm: DriverMode,
+    Buf: DmaTxBuffer,
+{
+    fn drop(&mut self) {
+        if !self.uhci.internal.channel.tx.is_done() {
+            self.uhci.internal.channel.tx.stop_transfer();
+            // TODO: Implement uhci drop
+
+            unsafe {
+                ManuallyDrop::drop(&mut self.uhci);
+                ManuallyDrop::drop(&mut self.dma_buf);
+            }
+        }
+    }
+}
+
+// Based on SpiDmaTransfer
+/// A structure representing a DMA transfer for UHCI/UART.
+///
+/// This structure holds references to the UHCI instance, DMA buffers, and
+/// transfer status.
+#[instability::unstable]
+pub struct UhciDmaRxTransfer<'d, Dm, Buf>
```

Comment:
Don't forget to add the `Deref` and `DerefMut` implementation for the buffer view